File: packages/wasmvm/wasmlib/ts/wasmlib/wasmtypes/scagentid.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

export const ScAgentIDNil: u8 = 0;
export const ScAgentIDAddress: u8 = 1;
export const ScAgentIDContract: u8 = 2;
export const ScAgentIDEthereum: u8 = 3;

import {panic} from '../sandbox';
import {
    addressFromBytes,
    addressFromString,
    addressToBytes,
    addressToString,
    ScAddress,
    ScAddressAlias,
    ScAddressEth,
    ScLengthAlias,
    ScLengthEd25519,
    ScLengthEth,
} from './scaddress';
import {chainIDFromBytes, ScChainIDLength} from './scchainid';
import {hnameFromBytes, hnameFromString, hnameToBytes, hnameToString, ScHname, ScHnameLength} from './schname';
import {concat, WasmDecoder, WasmEncoder} from './codec';
import {Proxy} from './proxy';

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

const nilAgentIDString = '-';

export class ScAgentID {
    kind: u8;
    _address: ScAddress;
    _hname: ScHname;
    eth: ScAddress;

    constructor(address: ScAddress, hname: ScHname) {
        this.kind = ScAgentIDContract;
        this._address = address;
        this._hname = hname;
        this.eth = new ScAddress();
    }

    public static forEthereum(chainAddress: ScAddress, ethAddress: ScAddress): ScAgentID {
        if (chainAddress.id[0] != ScAddressAlias) {
            panic("invalid eth AgentID: chain address");
        }
        if (ethAddress.id[0] != ScAddressEth) {
            panic("invalid eth AgentID: eth address");
        }
        const agentID = new ScAgentID(chainAddress, new ScHname(0));
        agentID.kind = ScAgentIDEthereum;
        agentID.eth = ethAddress;
        return agentID;
    }

    public static fromAddress(address: ScAddress): ScAgentID {
        const agentID = new ScAgentID(address, new ScHname(0));
        switch (address.id[0]) {
            case ScAddressAlias: {
                break;
            }
            case ScAddressEth: {
                panic("invalid eth AgentID: need chain address");
                break;
            }
            default: {
                agentID.kind = ScAgentIDAddress;
                break;
            }
        }
        return agentID;
    }

    public equals(other: ScAgentID): bool {
        return this._address.equals(other._address) &&
            this._hname.equals(other._hname);
    }

    public address(): ScAddress {
        return this._address;
    }

    public ethAddress(): ScAddress {
        return this.eth;
    }

    public hname(): ScHname {
        return this._hname;
    }

    public isAddress(): bool {
        return this.kind == ScAgentIDAddress;
    }

    public isContract(): bool {
        return this.kind == ScAgentIDContract;
    }

    // convert to byte array representation
    public toBytes(): Uint8Array {
        return agentIDToBytes(this);
    }

    // human-readable string representation
    public toString(): string {
        return agentIDToString(this);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function agentIDDecode(dec: WasmDecoder): ScAgentID {
    return agentIDFromBytes(dec.bytes());
}

export function agentIDEncode(enc: WasmEncoder, value: ScAgentID): void {
    enc.bytes(agentIDToBytes(value));
}

export function agentIDFromBytes(buf: Uint8Array | null): ScAgentID {
    if (buf === null || buf.length == 0) {
        const agentID = ScAgentID.fromAddress(addressFromBytes(null));
        agentID.kind = ScAgentIDNil;
        return agentID;
    }
    switch (buf[0]) {
        case ScAgentIDAddress: {
            buf = buf.subarray(1);
            if (buf.length != ScLengthAlias && buf.length != ScLengthEd25519) {
                panic('invalid AgentID length: address agentID');
            }
            return ScAgentID.fromAddress(addressFromBytes(buf));
        }
        case ScAgentIDContract: {
            buf = buf.subarray(1);
            if (buf.length != ScChainIDLength + ScHnameLength) {
                panic('invalid AgentID length: contract agentID');
            }
            const chainID = chainIDFromBytes(buf.subarray(0, ScChainIDLength));
            const hname = hnameFromBytes(buf.subarray(ScChainIDLength));
            return new ScAgentID(chainID.address(), hname);
        }
        case ScAgentIDEthereum: {
            buf = buf.subarray(1);
            if (buf.length != ScChainIDLength + ScLengthEth) {
                panic('invalid AgentID length: eth agentID');
            }
            const chainID = chainIDFromBytes(buf.subarray(0, ScChainIDLength));
            const ethAddress = addressFromBytes(buf.subarray(ScChainIDLength));
            return ScAgentID.forEthereum(chainID.address(), ethAddress);
        }
        case ScAgentIDNil:
            break;
        default: {
            panic('AgentIDFromBytes: invalid AgentID type');
            break;
        }
    }
    return agentIDFromBytes(null);
}

export function agentIDToBytes(value: ScAgentID): Uint8Array {
    let buf = new Uint8Array(1);
    buf[0] = value.kind;
    switch (value.kind) {
        case ScAgentIDAddress:
            return concat(buf, addressToBytes(value._address));
        case ScAgentIDContract: {
            buf = addressToBytes(value._address);
            buf[0] = value.kind;
            return concat(buf, hnameToBytes(value._hname));
        }
        case ScAgentIDEthereum: {
            buf = addressToBytes(value._address);
            buf[0] = value.kind;
            return concat(buf, addressToBytes(value.eth));
        }
        case ScAgentIDNil:
            return buf;
        default: {
            panic('AgentIDToBytes: invalid AgentID type');
            break;
        }
    }
    return buf;
}

export function agentIDFromString(value: string): ScAgentID {
    if (value == nilAgentIDString) {
        return agentIDFromBytes(null);
    }

    const parts = value.split('@');
    switch (parts.length) {
        case 1:
            return ScAgentID.fromAddress(addressFromString(parts[0]));
        case 2:
            if (!value.startsWith('0x')) {
                return new ScAgentID(addressFromString(parts[1]), hnameFromString(parts[0]));
            }
            return ScAgentID.forEthereum(addressFromString(parts[1]), addressFromString(parts[0]));
        default:
            panic('invalid AgentID string');
            return agentIDFromBytes(null);
    }
}

export function agentIDToString(value: ScAgentID): string {
    switch (value.kind) {
        case ScAgentIDAddress:
            return addressToString(value.address());
        case ScAgentIDContract: {
            return hnameToString(value.hname()) + '@' + addressToString(value.address());
        }
        case ScAgentIDEthereum:
            return addressToString(value.ethAddress()) + '@' + addressToString(value.address());
        case ScAgentIDNil:
            return nilAgentIDString;
        default: {
            panic('AgentIDToString: invalid AgentID type');
            return '';
        }
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableAgentID {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return agentIDToString(this.value());
    }

    value(): ScAgentID {
        return agentIDFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableAgentID extends ScImmutableAgentID {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: ScAgentID): void {
        this.proxy.set(agentIDToBytes(value));
    }
}

File: packages/wasmvm/wasmlib/go/wasmlib/wasmtypes/scagentid.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmtypes

import (
	"strings"
)

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

const (
	ScAgentIDNil      byte = 0
	ScAgentIDAddress  byte = 1
	ScAgentIDContract byte = 2
	ScAgentIDEthereum byte = 3
)

type ScAgentID struct {
	kind    byte
	address ScAddress
	hname   ScHname
}

const nilAgentIDString = "-"

func NewScAgentID(address ScAddress, hname ScHname) ScAgentID {
	return ScAgentID{kind: ScAgentIDContract, address: address, hname: hname}
}

func NewScAgentIDFromAddress(address ScAddress) ScAgentID {
	if address.id[0] == ScAddressAlias {
		return NewScAgentID(address, 0)
	}
	switch address.id[0] {
	case ScAddressAlias:
		return NewScAgentID(address, 0)
	case ScAddressEth:
		return ScAgentID{kind: ScAgentIDEthereum, address: address, hname: 0}
	default:
		return ScAgentID{kind: ScAgentIDAddress, address: address, hname: 0}
	}
}

func (o ScAgentID) Address() ScAddress {
	return o.address
}

func (o ScAgentID) Bytes() []byte {
	return AgentIDToBytes(o)
}

func (o ScAgentID) Hname() ScHname {
	return o.hname
}

func (o ScAgentID) IsAddress() bool {
	return o.kind == ScAgentIDAddress
}

func (o ScAgentID) IsContract() bool {
	return o.kind == ScAgentIDContract
}

func (o ScAgentID) String() string {
	return AgentIDToString(o)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

func AgentIDDecode(dec *WasmDecoder) ScAgentID {
	return AgentIDFromBytes(dec.Bytes())
}

func AgentIDEncode(enc *WasmEncoder, value ScAgentID) {
	enc.Bytes(AgentIDToBytes(value))
}

func AgentIDFromBytes(buf []byte) (a ScAgentID) {
	if len(buf) == 0 {
		return a
	}
	a.kind = buf[0]
	buf = buf[1:]
	switch a.kind {
	case ScAgentIDAddress:
		if len(buf) != ScLengthAlias && len(buf) != ScLengthEd25519 {
			panic("invalid AgentID length: address agentID")
		}
		a.address = AddressFromBytes(buf)
	case ScAgentIDContract:
		if len(buf) != ScChainIDLength+ScHnameLength {
			panic("invalid AgentID length: contract agentID")
		}
		a.address = ChainIDFromBytes(buf[:ScChainIDLength]).Address()
		a.hname = HnameFromBytes(buf[ScChainIDLength:])
	case ScAgentIDEthereum:
		if len(buf) != ScLengthETH {
			panic("invalid AgentID length: eth agentID")
		}
		a.address = AddressFromBytes(buf)
	case ScAgentIDNil:
		break
	default:
		panic("AgentIDFromBytes: invalid AgentID type")
	}
	return a
}

func AgentIDToBytes(value ScAgentID) []byte {
	buf := []byte{value.kind}
	switch value.kind {
	case ScAgentIDAddress:
		return append(buf, AddressToBytes(value.address)...)
	case ScAgentIDContract:
		buf = append(buf, AddressToBytes(value.address)[1:]...)
		return append(buf, HnameToBytes(value.hname)...)
	case ScAgentIDEthereum:
		return append(buf, AddressToBytes(value.address)...)
	case ScAgentIDNil:
		return buf
	default:
		panic("AgentIDToBytes: invalid AgentID type")
	}
}

func AgentIDFromString(value string) ScAgentID {
	if value == nilAgentIDString {
		return ScAgentID{}
	}

	parts := strings.Split(value, "@")
	switch len(parts) {
	case 1:
		return NewScAgentIDFromAddress(AddressFromString(parts[0]))
	case 2:
		return NewScAgentID(AddressFromString(parts[1]), HnameFromString(parts[0]))
	default:
		panic("invalid AgentID string")
	}
}

func AgentIDToString(value ScAgentID) string {
	switch value.kind {
	case ScAgentIDAddress:
		return AddressToString(value.Address())
	case ScAgentIDContract:
		return HnameToString(value.Hname()) + "@" + AddressToString(value.Address())
	case ScAgentIDEthereum:
		return AddressToString(value.Address())
	case ScAgentIDNil:
		// iscp.NilAgentID.String() returns "-" which means NilAgentID is "-"
		return nilAgentIDString
	default:
		panic("AgentIDToString: invalid AgentID type")
	}
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScImmutableAgentID struct {
	proxy Proxy
}

func NewScImmutableAgentID(proxy Proxy) ScImmutableAgentID {
	return ScImmutableAgentID{proxy: proxy}
}

func (o ScImmutableAgentID) Exists() bool {
	return o.proxy.Exists()
}

func (o ScImmutableAgentID) String() string {
	return AgentIDToString(o.Value())
}

func (o ScImmutableAgentID) Value() ScAgentID {
	return AgentIDFromBytes(o.proxy.Get())
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScMutableAgentID struct {
	ScImmutableAgentID
}

func NewScMutableAgentID(proxy Proxy) ScMutableAgentID {
	return ScMutableAgentID{ScImmutableAgentID{proxy: proxy}}
}

func (o ScMutableAgentID) Delete() {
	o.proxy.Delete()
}

func (o ScMutableAgentID) SetValue(value ScAgentID) {
	o.proxy.Set(AgentIDToBytes(value))
}
